### ISSUE TITLE
Type Inference Differs with Prefab Structs

### ISSUE THREAD

**Issue by MichaelBurge** [racket/typed-racket#685]
### What version of Racket are you using?

v6.12

### What program did you run?
```scheme
#lang typed/racket

(struct X ([ a : Integer ]) #:prefab)

(: *x* (Parameterof Integer))
(define *x* (make-parameter 0))

(: restore-x! (-> X Void))
(define (restore-x! x)
  (define x-a (X-a x))
  (*x* x-a)
  )
```
### What should have happened?

If the `#:prefab` is removed, the program typechecks and runs successfully. However, with `#:prefab`, the `x-a` variable has its type inferred as `Any`.

The program typechecks with a `(: x-a Integer)` type annotation, but the original program generates `x-a` in a macro which currently doesn't try to infer type information.

The behavior should be changed so that inference is the same for both opaque and prefab structs.

### If you got an error message, please include it here.

```
a.rkt:11:2: Type Checker: Wrong argument to parameter - expected Integer and got Any
  in: (*x* x-a)
```

**Comment by samth** [comment]
I think things can be improved here, but in general `#:prefab` structures are very different from normal structures, and can't be made to work the same. The easy way to see this is that `X-a` does not always return integers, because `(X-a #s(X "foo"))` produces a string.

**Comment by MichaelBurge** [comment]
It seems like general prefabs are a super-type of declared typed prefabs.

Could prefab literals could be protected by a contract or checked at compile-time to ensure that they are also a typed prefab?

**Comment by samth** [comment]
After further investigation, I have no idea how this could possibly happen. I'll keep looking.

**Comment by pnwamk** [comment]
After some _brief_ investigation with @samth, I believe this issue is related to (if not the same as) previous issues reported with prefab structs in Typed Racket (e.g. https://github.com/racket/typed-racket/issues/415). An overly-simplistic summary: they are currently very broken.

We believe we will have some time soon to look at what reasonable fixes might help.

**Comment by pnwamk** [comment]
Fixed by 7599aedf2c8ea354d1b4965ec129ea4dd85f8010.
